Library/src/Multiplayer/SignalRMsgPackEntitySerialiser.cpp

```diff
@@ -27,6 +27,178 @@ namespace csp::multiplayer
+			break;
+		case ReplicatedValueType::Vector2:
+		{
+			ValueType	= ItemComponentData::FLOAT_ARRAY;
+			auto VValue = Value.GetVector2();
+			NewValue	= signalr::value(std::vector {signalr::value(VValue.X), signalr::value(VValue.Y)});
+			break;
+		}
+		case ReplicatedValueType::Vector3:
+		{
+			ValueType	= ItemComponentData::FLOAT_ARRAY;
+			auto VValue = Value.GetVector3();
+			NewValue	= signalr::value(std::vector {signalr::value(VValue.X), signalr::value(VValue.Y), signalr::value(VValue.Z)});
+			break;
+		}
+		case ReplicatedValueType::Vector4:
+		{
+			ValueType	= ItemComponentData::FLOAT_ARRAY;
+			auto VValue = Value.GetVector4();
+			NewValue	= signalr::value(
+				   std::vector {signalr::value(VValue.X), signalr::value(VValue.Y), signalr::value(VValue.Z), signalr::value(VValue.W)});
+			break;
+		}
+		case ReplicatedValueType::Map:
+		{
+			ValueType = ItemComponentData::GENERIC_MAP;
+
+			// Creates a custom signalr data structure to hold an array of key-value pairs
+			const auto& Map = Value.GetMap();
+			auto Keys		= Map.Keys();
+
+			// Root of map which contains:
+			// 0 = Type information of the key
+			// 1 = Type information of value
+			// 2 = vector of key-value pairs
+			std::vector<signalr::value> MapRoot;
+			MapRoot.reserve(3);
+
+			// Array of key-value pairs
+			std::vector<signalr::value> MapArray;
+			MapArray.reserve(Keys->Size());
+
+			ItemComponentData KeyType	= ItemComponentData::UINT64;
+			ItemComponentData ValueType = ItemComponentData::UINT64;
+			bool First					= true;
+
+			for (size_t i = 0; i < Keys->Size(); ++i)
+			{
+				const ReplicatedValue& Key = (*Keys)[i];
+
+				auto KeyValue = ReplicatedValueToSignalRValue(Key);
+				auto Value	  = ReplicatedValueToSignalRValue(Map[Key]);
+
+				if (First)
+				{
+					First	  = false;
+					KeyType	  = KeyValue.first;
+					ValueType = Value.first;
+				}
+				else
+				{
+					// Ensure key and values are all the same type
+					assert(KeyType == KeyValue.first && "Key types dont match!");
+					assert(ValueType == Value.first && "Value types dont match!");
+				}
+
+				// Add key-value pair to the signalr value
+				MapArray.push_back(signalr::value(std::vector<signalr::value> {KeyValue.second, Value.second}));
+			}
+
+			CSP_DELETE(Keys);
+
+			MapRoot.push_back(signalr::value(KeyType));
+			MapRoot.push_back(signalr::value(ValueType));
+			MapRoot.push_back(MapArray);
+
+			// Create the root value with the array of key-value pairs
+			NewValue = signalr::value(MapRoot);
+
+			break;
+		}
+	}
+
+	return std::make_pair(ValueType, NewValue);
+}
+
+ReplicatedValue SignalRValueToReplicatedValue(ItemComponentData Type, const signalr::value& Value)
+{
+	switch (Type)
+	{
+		case ItemComponentData::BOOL:
+			return Value.as_bool();
+		case ItemComponentData::INT64:
+			return (Value.is_integer()) ? ReplicatedValue(Value.as_integer()) : ReplicatedValue((int64_t) Value.as_uinteger());
```

Comment:
c-style casting from a uint to an int may lead to value overflow/undefined behavior - worth doing something slightly more verbose/elaborate to avoid that?

Reply:
This is code moved from another function, I didn't write it in this PR.

Reply:
Happy to update if you think it'll be an issue. Are you imagining some kind of bounds checking?

Reply:
No, I wouldn't want to slow you down if this isn't a new issue. Mind logging an issue for it though?